src/pool/srv_target.c

```diff
@@ -827,13 +817,52 @@ ds_pool_hdl_hash_init(void)
 void
 ds_pool_hdl_hash_fini(void)
 {
-	/* Currently, we use "force" to purge all ds_pool_hdl objects. */
 	d_hash_table_destroy(pool_hdl_hash, true /* force */);
 }
 
+static int
+pool_hdl_delete_all_cb(d_list_t *link, void *arg)
+{
+	uuid_copy(arg, pool_hdl_obj(link)->sph_uuid);
+	return 1;
+}
+
+void
+ds_pool_hdl_delete_all(void)
+{
+	D_DEBUG(DB_MD, "deleting all pool handles\n");
+	D_ASSERT(dss_srv_shutting_down());
+
+	/*
+	 * The d_hash_table_traverse locking makes it impossible to delete or
+	 * even addref in the callback. Hence we traverse and delete one by one.
+	 */
+	for (;;) {
```

Comment:
I think we should add to warning comment above d_hash_rec_first for taking pointers without references, or better yet, just remove that API.

Reply:
This code has the same flaws doesn't it?  You're searching for and dropping references held or lost elsewhere so it's inherently only safe if any other threads that might hold those references have already been stopped.

Reply:
If this thread deletes an entry, then either a) the other thread won't find it, or b) the other thread has found it before this delete operation and is holding a reference, which prevents the entry from being freed. No?

Reply:
This function is searching for references and them dropping them isn't it?  So it's case b), the other thread looks up the pool, takes a reference, is working on the pool then this thread finds the reference, drops it frees the pool and the other thread then will do something invalid.  When we do this in dfuse it's safe because at that point all other threads have been shutdown, and actually the references held are held by the kernel which may or may not have been released depending on how the fuse connection was shutdown.

Perhaps I've misunderstood what this thread is doing, but I can't see why else you're traversing the hash table like this?

Reply:
This threads drops the reference held by the hash table during the delete operation. If the other thread has found the entry before this delete, it must have taken a reference, which should cause the delete _not_ to free the entry.

Reply:
So this is removing things from the hash table but not freeing them and it's doing it in a way that's safe against other threads doing essentially anything?  This would mean that no further lookups could succeed after this code had run but any other threads that held references could complete?

Reply:
Yes, as long as other threads are not buggy.

Reply:
P.S., essentially the difficulties might come from _some_ problem with the hash table API design, though I'm unable to identify clearly _what_---something doesn't feel right.

Reply:
Ok, I get it now.  I was wondering about dc_pool_get() although I note there's only one place we call it, and as the reference is held in the entry itself this should be fine anyway.

There is a comment above d_hash_rec_first() about safe usage but we could add a run-time check that it's only being used if the EPHEMERAL flag is set.